codec-http2/src/main/java/io/netty/handler/codec/http2/ReadOnlyHttp2Headers.java

```diff
@@ -0,0 +1,824 @@
+     *                     <a href="https://tools.ietf.org/html/rfc7540#section-8.1.2.1">pseudo headers</a>
+     *                     or {@code null} names/values.
+     * @param validateHeaders {@code true} will run validation on each header name/value pair to ensure protocol
+     *                        compliance.
+     * @return a new read only representation of headers used by clients.
+     */
+    public static ReadOnlyHttp2Headers clientHeaders(AsciiString method, AsciiString path, AsciiString scheme,
+                                                     AsciiString authority, AsciiString[] otherHeaders,
+                                                     boolean validateHeaders) {
+        return new ReadOnlyHttp2Headers(new AsciiString[] {
+                PseudoHeaderName.METHOD.value(), method, PseudoHeaderName.PATH.value(), path,
+                PseudoHeaderName.SCHEME.value(), scheme, PseudoHeaderName.AUTHORITY.value(), authority
+                }, otherHeaders, validateHeaders);
+    }
+
+    /**
+     * Create a new read only representation of headers used by servers.
+     * @param status The value for {@link PseudoHeaderName#STATUS}.
+     * @param otherHeaders A an array of key:value pairs. Must not contain any
+     *                     <a href="https://tools.ietf.org/html/rfc7540#section-8.1.2.1">pseudo headers</a>
+     *                     or {@code null} names/values.
+     * @param validateHeaders {@code true} will run validation on each header name/value pair to ensure protocol
+     *                        compliance.
+     * @return a new read only representation of headers used by servers.
+     */
+    public static ReadOnlyHttp2Headers serverHeaders(AsciiString status, AsciiString[] otherHeaders,
+                                                     boolean validateHeaders) {
+        return new ReadOnlyHttp2Headers(new AsciiString[] { PseudoHeaderName.STATUS.value(), status }, otherHeaders,
+                                        validateHeaders);
+    }
+
+    private ReadOnlyHttp2Headers(AsciiString[] pseudoHeaders, AsciiString[] otherHeaders, boolean validateHeaders) {
+        if ((pseudoHeaders.length & 1) != 0 || (otherHeaders.length & 1) != 0) {
+            throw new IllegalArgumentException("pseudoHeaders and otherHeaders must be arrays of [name, value] pairs");
+        }
+        if (validateHeaders) {
+            validateHeaders(pseudoHeaders, otherHeaders);
+        }
+        this.pseudoHeaders = pseudoHeaders;
+        this.otherHeaders = otherHeaders;
+    }
+
+    private static void validateHeaders(AsciiString[] pseudoHeaders, AsciiString[] otherHeaders) {
+        boolean seenNonPseudoHeader = false;
+        int i;
+        AsciiString name;
+        for (i = 0; i < pseudoHeaders.length; ++i) {
+            name = pseudoHeaders[i];
+            HTTP2_NAME_VALIDATOR.validateName(name);
+            if (!seenNonPseudoHeader && !name.isEmpty() && name.byteAt(0) != ':') {
```

Comment:
> iteration is not that expensive because it can skip the bounds check right?

Each individual access is be comparably cheaper than `byteAt()` but it is an `O(n)` iteration.

> Also, won't comparing ':' be an unnecessary upcast from byte to char?

Good catch. Let me be careful to avoid this.